### ISSUE TITLE
store rpc interface

### ISSUE THREAD

**Issue by joshrtay** [weo-edu/store#4]
how does koa (or other server side clients of the pipeline) invoke store methods directly?

**Comment by joshrtay** [comment]
seems like we should probably try to write an nsq rpc implementation

**Comment by joshrtay** [comment]
Proposal 1:

1 rpc topic, 1 callback topic

Proposal 2:

1 rpc topic, uniqe topic per callback

Proposal 3:

1 topic per procedure, unique topic per callback

I feel like you could get better guarantees on delivery with more topics but i'm not quite sure.

**Comment by joshrtay** [comment]
http://nsq.io/overview/faq.html its mentioned int he faq

**Comment by joshrtay** [comment]
i think proposal 2 is the best.

**Comment by ashaffer** [comment]
Hmm...i'm not sure why we would actually want to do this for RPC.  It seems like koa can simply require and call the store's exported methods directly, without any loss of generality or flexibility since data-retrieval is side-effect free.

I'm not sure I see the advantage of multiple topics for callbacks either.  I would think the delivery guarantees would be the same too.  Single topic for store -> koa communication seems like the most straightforward concept to me.

**Comment by ashaffer** [comment]
We also have to make each koa server it's own channel, I just realized.  Every koa server has to process every message from the stores, because we don't know in advance which messages are callbacks to a particular event, or which socket.io clients are connected to which servers.

EDIT: I think this also means we could theoretically ditch sticky routing, though I think socket.io kind of relies on it, so maybe we would want to switch to something lower-level like sockjs.  nvm sockjs requires sticky sessions too.

**Comment by joshrtay** [comment]
the reason we would want to use nsq for rpc is that i was thinking that the stores lived in a separate process from koa and nsq would give you a way to communicate with the stores and give you roud robin load balancing.

**Comment by joshrtay** [comment]
still don't know what we do here. nsq mentions on their FAQ page that people who are interested in rpc should reach out to them. they seem pretty responsive.

**Comment by ashaffer** [comment]
I don't think we want to do RPC for gets.  The reason we're doing it for writes is so that we can persist events in a canonical way, and distribute them around because they process slowly and have potentially cascading effects.  Reads don't have those properties.  Reads will already be load balanced by whatever clustering / read replication we do with our data stores.  The client-library code for reads is thin and has no real reason to be load-balanced.  Adding RPC to that would just introduce unnecessary latency and complexity without any real benefit IMO.
